src/ws/cockpitauth.c

```diff
@@ -550,8 +551,16 @@ create_creds_for_spawn_authenticated (CockpitAuth *self,
                              COCKPIT_CRED_CSRF_TOKEN, csrf_token,
                              NULL);
 
-  g_free (csrf_token);
+  if (!cockpit_json_get_object (results, "login-data", NULL, &login_data))
+    {
+      g_warning ("received bad login-data from %s", sl->command);
+      login_data = NULL;
+    }
 
+  if (login_data)
+    cockpit_creds_set_login_data (creds, login_data);
```

Comment:
An alternative would be to return the JsonObject from cockpit_auth_login_finish() and pass it to send_login_response(). We don't need to keep this data loaded in cockpit-ws, and thus it doesn't really need to be part of CockpitCreds.